vortex-cli/src/browse/app.rs

```diff
@@ -0,0 +1,215 @@
+use std::ops::Range;
+use std::os::unix::fs::FileExt;
+use std::path::Path;
+use std::sync::Arc;
+
+use ratatui::widgets::ListState;
+use vortex::buffer::{Alignment, ByteBuffer, ByteBufferMut};
+use vortex::dtype::{DType, Field, Nullability, StructDType};
+use vortex::error::{VortexExpect, VortexResult};
+use vortex::file::{
+    ExecutionMode, FileLayout, Segment, SplitBy, VortexOpenOptions, CHUNKED_LAYOUT_ID,
+    COLUMNAR_LAYOUT_ID, FLAT_LAYOUT_ID,
+};
+use vortex::io::TokioFile;
+use vortex::sampling_compressor::ALL_ENCODINGS_CONTEXT;
+use vortex::stats::stats_from_bitset_bytes;
+use vortex_layout::segments::SegmentId;
+use vortex_layout::{LayoutData, LayoutEncodingRef};
+// Add a shared Tokio Runtime for use in the app.
+
+#[derive(Default, Copy, Clone, Eq, PartialEq)]
+pub enum Tab {
+    /// The layout tree browser.
+    #[default]
+    Layout,
+    /// The encoding tree viewer
+    Encodings,
+    // TODO(aduffy): SQL query page powered by DF
+    // Query,
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Encoding {
+    Flat,
+    Chunked,
+    Columnar,
+    Unknown,
+}
+
+impl From<u16> for Encoding {
+    fn from(value: u16) -> Self {
+        if value == FLAT_LAYOUT_ID.0 {
+            Encoding::Flat
+        } else if value == CHUNKED_LAYOUT_ID.0 {
+            Encoding::Chunked
+        } else if value == COLUMNAR_LAYOUT_ID.0 {
+            Encoding::Columnar
+        } else {
+            Encoding::Unknown
+        }
+    }
+}
+
+/// A pointer into the `Layout` hierarchy that can be advanced.
+///
+/// The pointer wraps an InitialRead.
+pub struct LayoutCursor {
```

Comment:
Is this somewhat similar to ArrayVisitor?

Reply:
Somewhat? It's specifically for layout hierarchy, and just gives us an owned pointer that we can mutate to move up/down the tree